library/core/src/primitive_docs.rs

```diff
@@ -1833,6 +1833,8 @@ mod prim_ref {}
 /// - If `T` is guaranteed to be subject to the [null pointer
 ///   optimization](option/index.html#representation), and `E` is an enum satisfying the following
 ///   requirements, then `T` and `E` are ABI-compatible. Such an enum `E` is called "option-like".
+///   - The enum `E` uses the [`Rust` representation], and is not modified by the `align` or
+///     `packed` representation modifiers.
 ///   - The enum `E` has exactly two variants.
 ///   - One variant has exactly one field, of type `T`.
 ///   - All fields of the other variant are zero-sized with 1-byte alignment.
```

Comment:
Reviewing more closely, I feel guaranteeing e.g. `repr(align(2))` won't change the layout if `T` is 2-aligned would be inconsistent with this clause.

We could potentially raise this alignment to whatever the alignment of `T` is, for example. I don't want to alter the guarantee around the 1ZST case, however, as it places a number of annoying requirements on the implementation. Being able to special-case specifically on 1ZSTs simplifies a lot of things. I would be okay with guaranteeing that `repr(align(1))` doesn't change anything, though, because it doesn't. There's also some inclarity about certain special cases around ZSTs already when they combine with other repr attributes like `repr(C)`.